[pytorch][caffe2] Add support for quantized slice conversion

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30679 [onnx][quantization] Add JIT pass to insert permutes for conv ops
* **#30498 [pytorch][caffe2] Add support for quantized slice conversion**

Summary:
Updated Int8SliceOp to accept dim, start and end index similar to Pytorch.

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_slice

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18740519](https://our.internmc.facebook.com/intern/diff/D18740519)

supriyar committed:
[pytorch][caffe2] Add support for quantized slice conversion
Summary:
Updated Int8SliceOp to accept dim, start and end index similar to Pytorch.

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_slice

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

houseroad commented on caffe2/operators/quantized/int8_slice_op.cc:
```diff
@@ -38,6 +38,11 @@ negative (end index -1 means up to and including the last element).
+    .Arg(
+        "dim",
+        "(Optional) The dimension to slice over. If specified start_idx and end_idx should also be given and it takes precedence over starts and ends")
+    .Arg("start_idx", "(Optional) The dimension to start slice from")
```
Mention the default value here?

houseroad commented on caffe2/operators/quantized/int8_slice_op.cc:
```diff
@@ -38,6 +38,11 @@ negative (end index -1 means up to and including the last element).
+        "dim",
+        "(Optional) The dimension to slice over. If specified start_idx and end_idx should also be given and it takes precedence over starts and ends")
+    .Arg("start_idx", "(Optional) The dimension to start slice from")
+    .Arg("end_idx", "(Optional) The dimension to end the slice")
```
ditto

houseroad commented on caffe2/operators/quantized/int8_slice_op.h:
```diff
@@ -35,6 +35,18 @@ class Int8SliceOp final : public SliceOp<CPUContext> {
+        if (HasArgument("dim")) {
+          auto dim = this->template GetSingleArgument<int>("dim", 0);
+          auto start = this->template GetSingleArgument<int64_t>("start_idx", 0);
+          auto end = this->template GetSingleArgument<int64_t>("end_idx", 0);
```
Shouldn't this be -1?

houseroad commented on torch/onnx/symbolic_caffe2.py:
```diff
@@ -192,3 +192,30 @@ def reshape(g, input, shape):
+    if input not in sym_help._quantized_ops:
+        from torch.onnx.symbolic_opset9 import slice
+        return slice(g, input, dim, start, end, step)
+
```
error out if step != 1?

houseroad reviewed (changes requested):
Please check my inline comments

supriyar committed:
Update on "[pytorch][caffe2] Add support for quantized slice conversion"


Summary:
Updated Int8SliceOp to accept dim, start and end index similar to Pytorch.

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_slice

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[pytorch][caffe2] Add support for quantized slice conversion"


Summary:
Updated Int8SliceOp to accept dim, start and end index similar to Pytorch.

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_slice

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

houseroad commented on caffe2/operators/quantized/int8_slice_op.h:
```diff
@@ -35,6 +35,18 @@ class Int8SliceOp final : public SliceOp<CPUContext> {
         CAFFE_ENFORCE_EQ(starts_.size(), ends_.size());
 
+        if (HasArgument("dim")) {
+          auto dim = this->template GetSingleArgument<int>("dim", 0);
```
Nit: Check starts and ends don't exist, otherwise, at least print out some warning.

supriyar replied:
Do you mean start_idx and end_idx? starts and ends are expected to exist for this op, I didn't change that behavior.

houseroad replied:
No, I just meant starts and ends here. After latest changes, I think no need to check any more.

houseroad commented on torch/onnx/symbolic_caffe2.py:
```diff
@@ -187,3 +185,32 @@ def reshape(g, input, shape):
+    dim = sym_help._parse_arg(dim, 'i')
+
+    # Dummy values as C2 op expects them to be passed.
+    start_list = [0]
```
If we don't provide starts and ends, will that cause crash or not? If so, we probably want to relax the constrains a bit.

supriyar replied:
Yes, it will crash. I didn't want to change existing behavior for existing ops so didn't modify that. @houseroad do you think I should?

houseroad replied:
I think it's fine, it should backward compatible. Let's move the ENFORCE on starts/ends to the else branch of checking existence of `dim`.

houseroad reviewed (commented):
Looks good, two more nits

supriyar committed:
Update on "[pytorch][caffe2] Add support for quantized slice conversion"


Summary:
Updated Int8SliceOp to accept dim, start and end index similar to Pytorch.

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_slice

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18740519](https://our.internmc.facebook.com/intern/diff/D18740519)

[ghstack-poisoned]

houseroad reviewed (approved):
Shipit!

supriyar committed:
Update on "[pytorch][caffe2] Add support for quantized slice conversion"


Summary:
Updated Int8SliceOp to accept dim, start and end index similar to Pytorch.

Test Plan:
python test/onnx/test_pytorch_onnx_caffe2_quantized.py TestQuantizedOps.test_slice

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D18740519](https://our.internmc.facebook.com/intern/diff/D18740519)

[ghstack-poisoned]
